### PR TITLE
ENH: Refactoring of if-statements

### PR DESCRIPTION
**Pull request by Anselmoo**
#### Description

Refactoring of if statements in `confidence.py`, `minimizer.py`, `model.py`, and `parameter.py`

###### Type of Changes

- [ ] Bug fix
- [ ] New feature
- [x] Refactoring / maintenance
- [ ] Documentation / examples

###### Tested on

```shell
Python: 3.8.12 (default, Nov 11 2021, 07:35:10) 
[Clang 13.0.0 (clang-1300.0.29.3)]

lmfit: 1.0.3, scipy: 1.7.2, numpy: 1.21.4, asteval: 0.9.25, uncertainties: 3.1.6
```
###### Verification 
Have you

- [x] included docstrings that follow PEP 257?

- [ ] referenced existing Issue and/or provided relevant link to mailing list?

- [x] verified that existing tests pass locally?

- [ ] verified that the documentation builds locally?

- [ ] squashed/minimized your commits and written descriptive commit messages?

- [ ] added or updated existing tests to cover the changes?
- [ ] updated the documentation and/or added an entry to the release notes (doc/whatsnew.rst)?
- [ ] added an example?

### CODE REVIEW

**Comment by newville** [comment]
@Anselmoo I think we all appreciate the interest and effort.  I also think that an unsolicited pull request that modifies working code to match your own style preference might not be the best introduction one can make. 

@reneeotten any thoughts on this?

**Comment by reneeotten** [comment]
> @Anselmoo I think we all appreciate the interest and effort. I also think that an unsolicited pull request that modifies working code to match your own style preference might not be the best introduction one can make.
> 
> @reneeotten any thoughts on this?

@newville @Anselmoo some of rewrites make sense and I'd be in favor of using them, others I don't reaaly like from cursory look. I will take a closer look hopefully in the next few days and leave some review comments on the proposed changes.

**Comment by newville** [comment]
@reneeotten thanks -- that was my impression too.  I guess that means that we should do a line-by-line review of this over the next week or so (I'm not going to rush on this!) and hope that @Anselmoo is amenable to making the requested changes.

**Comment by Anselmoo** [comment]
> @reneeotten thanks -- that was my impression too. I guess that means that we should do a line-by-line review of this over the next week or so (I'm not going to rush on this!) and hope that @Anselmoo is amenable to making the requested changes.

@newville @reneeotten this sounds good

File: lmfit/model.py [hunk 1]
```diff
@@ -460,22 +460,19 @@ def _parse_params(self):
...
         #   kw_args: dict of keyword arguments with default values
         #   keywords_:  name of **kws argument or None
         # 1. limited support for asteval functions as the model functions:
         if hasattr(self.func, 'argnames') and hasattr(self.func, 'kwargs'):
             pos_args = self.func.argnames[:]
-            keywords_ = None
-            kw_args = {}
             for name, defval in self.func.kwargs:
                 kw_args[name] = defval
-        # 2. modern, best-practice approach: use inspect.signature
```
Comment (reneeotten) [inline lmfit/model.py:474]: please leave this comment

File: lmfit/minimizer.py [hunk 2]
```diff
@@ -591,16 +589,16 @@ def __residual(self, fvars, apply_bounds_transformation=True):
...
+        if not self._abort:
             return _nan_policy(np.asarray(out).ravel(),
                                nan_policy=self.nan_policy)
 
+        self.result.residual = out
+        self.result.aborted = True
+        self.result.message = "Fit aborted by user callback. Could not estimate error-bars."
+        self.result.success = False
+        raise AbortFitException("fit aborted by user.")
+
```
Comment (reneeotten) [inline lmfit/minimizer.py:601]: please revert this block to the original; there is no benefit of doing this with respect to speed / readability

File: lmfit/minimizer.py [hunk 3]
```diff
@@ -647,16 +645,17 @@ def penalty(self, fvars):
...
         """
-        if self.result.method in ['brute', 'shgo', 'dual_annealing']:
-            apply_bounds_transformation = False
-        else:
-            apply_bounds_transformation = True
+        apply_bounds_transformation = self.result.method not in [
+            'brute',
+            'shgo',
+            'dual_annealing',
+        ]
```
Comment (reneeotten) [inline lmfit/minimizer.py:652]: please revert this, this reverse logic makes it much less readable at first glance...

File: lmfit/minimizer.py [hunk 4]
```diff
@@ -647,16 +645,17 @@ def penalty(self, fvars):
...
 
         r = self.__residual(fvars, apply_bounds_transformation)
         if isinstance(r, np.ndarray) and r.size > 1:
             r = self.reduce_fcn(r)
-            if isinstance(r, np.ndarray) and r.size > 1:
-                r = r.sum()
+        if isinstance(r, np.ndarray) and r.size > 1:
+            r = r.sum()
```
Comment (reneeotten) [inline lmfit/minimizer.py:658]: I don't see a benefit here and am sure the changed code will work as well; to me the flow is visually easier to follow in the old code; the second check/manipulation will only happen if the first criteria is met. I'd prefer to revert.

File: lmfit/minimizer.py [hunk 5]
```diff
@@ -1365,13 +1362,12 @@ def emcee(self, params=None, steps=1000, nwalkers=100, burn=0, thin=1,
...
-            if param.vary:
-                var_arr[i] = param.value
-                i += 1
-            else:
+            if not param.vary:
                 # don't want to append bounds if they're not being varied.
                 continue
 
+            var_arr[i] = param.value
+            i += 1
```
Comment (reneeotten) [inline lmfit/minimizer.py:1370]: please revert, no benefit in readability / shorter code

File: lmfit/minimizer.py [hunk 6]
```diff
@@ -1497,34 +1493,38 @@ def emcee(self, params=None, steps=1000, nwalkers=100, burn=0, thin=1,
...
 
         # If uncertainty was automatically estimated, weight the residual properly
-        if (not is_weighted) and (result.residual.size > 1):
-            if '__lnsigma' in params:
-                result.residual = result.residual/np.exp(params['__lnsigma'].value)
+        if (
+            (not is_weighted)
+            and (result.residual.size > 1)
+            and '__lnsigma' in params
+        ):
```
Comment (reneeotten) [inline lmfit/minimizer.py:1500]: I am okay with folding the second `if`-statement in here, but don't like the formatting - I don't think all these parenthesis are actually needed or (I know the were there already...).
Can this not just become:
```
if not is_weighted and result.residual.size > 1 and '__lnsigma' in params:
```

File: lmfit/minimizer.py [hunk 7]
```diff
@@ -1497,34 +1493,38 @@ def emcee(self, params=None, steps=1000, nwalkers=100, burn=0, thin=1,
...
 
-        if auto_pool is not None:
-            auto_pool.terminate()
+        # assumes that residual is properly weighted, avoid overflowing np.exp()
+        result.chisqr = np.exp(min(650, _neg2_log_likel))
 
-        return result
+        result.redchi = result.chisqr / result.nfree
+        result.aic = _neg2_log_likel + 2 * result.nvarys
+        result.bic = _neg2_log_likel + np.log(result.ndata) * result.nvarys
```
Comment (reneeotten) [inline lmfit/minimizer.py:1527]: I don't follow this as the diff is hard to read; what is the actual change here? Why do we need a new function `_extracted_from_emcee_201` - unless there is a convincing argument to do this, I am not really inclined to accept this.

**Review by reneeotten — CHANGES_REQUESTED** [review]
@Anselmoo sorry for the delay... I went looked through your proposed changes and left comments for the ones I don't agree with and/or have a question. Please take a look and once you make the changes / comment here, perhaps @newville can glance over it as well and we can wrap this up.

File: lmfit/model.py (hunk 1 above)
Comment (Anselmoo) [reply lmfit/model.py:474]: Agree

**Comment by reneeotten** [comment]
@Anselmoo a few more things and then please squash your commits so there is only one left

File: lmfit/minimizer.py [hunk 8]
```diff
@@ -1497,16 +1492,13 @@ def emcee(self, params=None, steps=1000, nwalkers=100, burn=0, thin=1,
                                       handle_inf=False)
 
         # If uncertainty was automatically estimated, weight the residual properly
-        if (not is_weighted) and (result.residual.size > 1):
-            if '__lnsigma' in params:
-                result.residual = result.residual/np.exp(params['__lnsigma'].value)
+        if not is_weighted and result.residual.size > 1 and '__lnsigma' in params:
+            result.residual = result.residual/np.exp(params['__lnsigma'].value)
```
[the remark this answers is not in the document]
Comment (Anselmoo) [reply lmfit/minimizer.py:1496]: That's an excellent point

File: lmfit/minimizer.py [hunk 9]
```diff
@@ -1497,16 +1492,13 @@ def emcee(self, params=None, steps=1000, nwalkers=100, burn=0, thin=1,
...
 
         # Calculate statistics for the two standard cases:
         if isinstance(result.residual, np.ndarray) or (float_behavior == 'chi2'):
             result._calculate_statistics()
 
-        # Handle special case unique to emcee:
-        # This should eventually be moved into result._calculate_statistics.
```
[the remark this answers is not in the document]
Comment (Anselmoo) [reply lmfit/minimizer.py:1509]: That was a mistake; good point

**Comment by Anselmoo** [comment]
> @Anselmoo a few more things and then please squash your commits so there is only one left

I hope my combination of local _soft reset_, _squash_, and _push_, is working as expected.

**Review by reneeotten — APPROVED** [review]

**Comment by reneeotten** [comment]
thanks for taking care of the review comments/changes @Anselmoo! This looks good to me - @newville can you glance over this to make sure you're also happy with the proposed changes?

**Comment by newville** [comment]
@Anselmoo @reneeotten My only concern is that the change at `Model.py` line 290 should be reverted to `if len(opts) > 0:`.  All the other changes are cosmetic, but that one changes the logic.

**Comment by reneeotten** [comment]
> @Anselmoo @reneeotten My only concern is that the change at `Model.py` line 290 should be reverted to `if len(opts) > 0:`. All the other changes are cosmetic, but that one changes the logic.

@newville: I don't think that statement is correct, there is no change in the logic: `if len(opts) > 0` just tests if the list exists (i.e., has at least one element), that's the same as what happens with `if opts` - right? According to `pep8` the latter is actually preferred: "For sequences, (strings, lists, tuples), use the fact that empty sequences are false".

**Comment by newville** [comment]
@reneeotten part of the question is where you want a possible error to occur and what the message to say.  Please leave `if len(opts) > 0`.  

Really, I appreciate the willingness to help.  Please do not re-format working code to match a style guide.  If the two forms are equivalent, what is gained by the change?  If the two forms are not (say, someone adds some code and `opts` is no longer a sequence), then the code is broken.

**Comment by Anselmoo** [comment]
@newville @reneeotten thx for your review

>If the two forms are equivalent, what is gained by the change?

Concerning this argument, the two forms (`if len(opts) > 0` vs. `if opts > 0`) are not equivalent, because `if len(opts) > 0` is redundant if statements. Just comparing the list would be enough. 

```python
>>> [ ] == True
False
```

In case, that `opts` can be also become a just an `int` or `float`,  a `raise`-statement would make sense?

But I got your point @newville, this was just for clarificaion.

**Comment by Anselmoo** [comment]
@reneeotten any update?

**Comment by newville** [comment]
@Anselmoo The ball is still in your court to address the existing reviews.   Revert Model line 292.  Yes, that logic change is holding up merging this.

**Comment by Anselmoo** [comment]
> @Anselmoo The ball is still in your court to address the existing reviews. Revert Model line 292. Yes, that logic change is holding up merging this.

Thx for feedback @newville

**Comment by Anselmoo** [comment]
I can remove the rebase, however it looks like that _five_ test will be failed in any case if this PR will be merged with the master branch.

@newville @reneeotten are there any plans to fix this in another PR or should be this fixed in this, too?

Most of the fails are linked to numeric issues [https://github.com/lmfit/lmfit-py/pull/763/checks?check_run_id=5048292287](https://github.com/lmfit/lmfit-py/pull/763/checks?check_run_id=5048292287) after specifying the `requirements`?

**Comment by reneeotten** [comment]
> I can remove the rebase, however it looks like that _five_ test will be failed in any case if this PR will be merged with the master branch.

Yes, squash your commits and rebase with `master` so that you have only one commit left. There are some tests that fail indeed and this started with the latest NumPy version but only on Linux. I don't really understand where the problem originates... Of course if you know how to fix it issue (without just relaxing the criteria and/or removing the tests of course) that would be great, but it's not a prerequisite for this PR.

**Comment by reneeotten** [comment]
thanks @Anselmoo

**Comment by newville** [comment]
@Anselmoo @reneeotten thanks!

**Pull request merged**